File: 05/main.py
from __future__ import annotations
from collections import defaultdict
from typing import DefaultDict, Iterator, List, NamedTuple
from parse import parse


class Point(NamedTuple):
    x: int
    y: int

    def __add__(self, other: Point):
        return Point(self.x + other.x, self.y + other.y)


class Line(NamedTuple):
    start: Point
    end: Point

    def draw(self) -> Iterator[Point]:
        dx: int = find_delta(self.start.x, self.end.x)
        dy: int = find_delta(self.start.y, self.end.y)
        delta: Point = Point(dx, dy)
        point: Point = self.start
        while point != self.end:
            yield point
            point += delta
        yield self.end


def find_delta(start: int, end: int) -> int:
    if start < end:
        return 1
    if start > end:
        return -1
    return 0


def parse_line(line: str) -> Line:
    p: dict = parse("{sx:d},{sy:d} -> {ex:d},{ey:d}", line).named
    return Line(Point(p["sx"], p["sy"]), Point(p["ex"], p["ey"]))


def read_puzzle_input(filename: str) -> List[Line]:
    with open(filename) as stream:
        return [parse_line(line) for line in stream.read().strip().split("\n")]


def check_if_orthogonal(line: Line):
    return line.start.x == line.end.x or line.start.y == line.end.y


def count_overlapping_points(lines: List[Line]):
    board: DefaultDict[Point, int] = defaultdict(int)
    for line in lines:
        for point in line.draw():
            board[point] += 1
    return len([p for p, c in board.items() if c > 1])


if __name__ == "__main__":
    lines: List[Line] = read_puzzle_input("input.txt")
    orthogonal_lines: List[Line] = [line for line in lines if check_if_orthogonal(line)]
    print(count_overlapping_points(orthogonal_lines))
    print(count_overlapping_points(lines))

File: 05/main.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Point struct {
	x int
	y int
}

func (p *Point) add(other Point) Point {
	return Point{p.x + other.x, p.y + other.y}
}

type Line struct {
	start Point
	end   Point
}

func findDelta(a int, b int) int {
	if a < b {
		return 1
	} else if a > b {
		return -1
	}
	return 0
}

func (l *Line) draw() *[]Point {
	points := []Point{l.start}
	delta := Point{findDelta(l.start.x, l.end.x), findDelta(l.start.y, l.end.y)}
	point := l.start
	for point != l.end {
		point = point.add(delta)
		points = append(points, point)
	}
	return &points
}

func (l *Line) isOrthogonal() bool {
	return l.start.x == l.end.x || l.start.y == l.end.y
}

func countOverlappingPoints(lines *[]Line) int {
	pointCounts := map[Point]int{}
	for _, line := range *lines {
		for _, point := range *line.draw() {
			pointCounts[point] += 1
		}
	}
	count := 0
	for _, pc := range pointCounts {
		if pc > 1 {
			count += 1
		}
	}
	return count
}

func getOrthogonalLines(lines *[]Line) *[]Line {
	orthogonalLines := []Line{}
	for _, line := range *lines {
		if line.isOrthogonal() {
			orthogonalLines = append(orthogonalLines, line)
		}
	}
	return &orthogonalLines
}

func parsePoint(pointString string) Point {
	coords := []int{}
	for _, coordString := range strings.Split(pointString, ",") {
		coord, err := strconv.Atoi(coordString)
		if err != nil {
			panic(err)
		}
		coords = append(coords, coord)
	}
	return Point{coords[0], coords[1]}
}

func readPuzzleInput(filePath string) *[]Line {
	lines := []Line{}
	data, err := os.ReadFile(filePath)
	if err != nil {
		panic(err)
	}
	for _, lineString := range strings.Split(strings.TrimSpace(string(data)), "\n") {
		parts := strings.Split(lineString, " -> ")
		start := parsePoint(parts[0])
		end := parsePoint(parts[1])
		lines = append(lines, Line{start, end})
	}
	return &lines
}

func main() {
	lines := readPuzzleInput("input.txt")
	fmt.Println(countOverlappingPoints(getOrthogonalLines(lines)))
	fmt.Println(countOverlappingPoints(lines))
}
